### ISSUE TITLE
Changing default `n_chains` on GPUs and `n_discard_per_chain` to something more reasonable

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#1793]
Right now:
 - `n_chains_per_rank` defaults to `16`.
 - `n_discard_per_chain` defaults to `n_samples//10`

The former is very bad if we are using GPUs, the latter is always bad as was discussed in #1034 with @gcarleo and @wdphy16 .

I propose to change:
 - `n_chains_per_rank` defaults to `16` on CPUs and other backends, `512` on GPUs. (it would be nice to know how 'big a gpu is' but I think this is a reasonable number) 
 - `n_discard_per_chain` to `10`, which in practice seems a reasonable number (and much more reasonable than the current value). It's still a bit high for GPUs and lower values work in practice, but with the new timing utilities it should be easier to spot.

What do you think? @gcarleo @jwnys @inailuig @wdphy16 @alleSini99

**Comment by jwnys** [comment]
Lowering the n_discard_per_chain sounds like a great idea (I default to 4). It sounds a bit strange to me, though, to change the number of chains based on the backend (this is what will effectively happen, right?). I would rather take the number of chains constant. If you run on CPU, the number of ranks will be high, and so the number of chains per rank low, while the opposite then happens on GPU.

**Comment by gcarleo** [comment]
Yes I would use a constant number of discarded (not a fraction) but I wouldn't be in favor of having backend dependent defaults,this might be dangerous and give different results unexpectedly

**Comment by alleSini99** [comment]
I super agree about the constant and small number of discards per chain. I would keep the number of chains the same on the different backends, though, since the opposite could generate unexpected behaviours.

**Comment by PhilipVinc** [comment]
Ok, so it seems that there is agreement on a lower `n_discard_per_chain`. Do we all agree on `10` or is this too large?

About changing `n_chains_per_rank` dynamically...
 - I am proposing to set the default value of `n_chains_per_rank` based on the backend (@jwnys , not the `n_chains`) 

Right now the default is 16 chains per rank, which is ok for CPUs (even a bit high in some cases when running under MPI) but completely not ok for GPUs.

The reason I'd like to set this is because the current default is completely useless on GPUs, and I see the defaults as something that newcomers can learn from.

What do you fear might break if we change the number of chains dynamically?
